ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,173 @@
+         && contains_substring record.API.message_body group_uuid
+     )
+
+(** If it is breach and no alerts existed, generate one, 
+    If it is not breach and alerts existed, dismiss the existed alert *)
+let update_vm_placement_alert_with_group ~__context ~group ~records =
+  try
+    let breach_on_host = check_breach ~__context ~group in
+    let alert_records = filter_alert_with_group ~__context ~group ~records in
+    debug "The length of group (UUID: %s) is: %d"
+      (Db.VM_group.get_uuid ~__context ~self:group)
+      (List.length alert_records) ;
+    match (alert_records, breach_on_host) with
+    | [], Some host ->
+        report_anti_affinity_alert ~__context ~group ~host
+    | records, None ->
+        List.iter
+          (fun (ref, _) ->
+            Helpers.call_api_functions ~__context (fun rpc session_id ->
+                Client.Client.Message.destroy ~rpc ~session_id ~self:ref
+            )
+          )
+          records
+    | _, _ ->
+        ()
+  with _ -> warn "Update VM anti-affinity alert failed"
+
+let update_vm_placement_alert_with_vm ~__context ~vm =
+  let group_opt =
+    List.find_opt
+      (fun g -> Db.VM_group.get_placement ~__context ~self:g = `anti_affinity)
+      (Db.VM.get_groups ~__context ~self:vm)
+  in
+  match group_opt with
+  | None ->
+      ()
+  | Some group ->
+      update_vm_placement_alert_with_group ~__context ~group
+        ~records:(get_all_records ~__context)
+
+let remove_vm_placement_alert_with_group ~__context ~group ~records =
+  filter_alert_with_group ~__context ~group ~records
+  |> List.iter (fun (ref, _) ->
+         Helpers.call_api_functions ~__context (fun rpc session_id ->
+             Client.Client.Message.destroy ~rpc ~session_id ~self:ref
+         )
+     )
+
+let update_vm_placement_alert ~__context ~groups =
+  let records = get_all_records ~__context in
+  groups
+  |> List.filter (fun g ->
+         Db.VM_group.get_placement ~__context ~self:g = `anti_affinity
+     )
+  |> List.iter (fun group ->
+         update_vm_placement_alert_with_group ~__context ~group ~records
+     )
+
+let remove_vm_placement_alert ~__context ~groups =
+  let records = get_all_records ~__context in
+  groups
+  |> List.filter (fun g ->
+         Db.VM_group.get_placement ~__context ~self:g = `anti_affinity
+     )
+  |> List.iter (fun group ->
+         remove_vm_placement_alert_with_group ~__context ~group ~records
+     )
+
+let update_vm_placement_alert_on_features_change ~__context ~old_restrictions
+    ~new_restrictions =
+  try
+    let feature_name = Features.name_of_feature Features.VM_anti_affinity in
+    let feature = "restrict_" ^ String.lowercase_ascii feature_name in
+    let old_feature_restriction = List.assoc_opt feature old_restrictions in
+    let new_feature_restriction = List.assoc_opt feature new_restrictions in
+    match (old_feature_restriction, new_feature_restriction) with
+    | Some old_flag, Some new_flag when old_flag <> new_flag -> (
```

Comment:
I think we'd better also handle `None` for `old_feature_restriction` and `new_feature_restriction`, for example, in case the feature is removed later.

Reply:
I have updated and add the logic to handle this corner case